### PR TITLE
Onboard endnode

### PR DESCRIPTION
**Pull request by yongpingchen**

### CODE REVIEW

File: api.go [hunk 1]
```diff
@@ -189,6 +189,22 @@ type OnboardByOwnerRequest struct {
 	LayoutPosition string `json:"layoutPosition,omitempty"` // pattern: GATEWAY|STANDALONE|ENDNODE, STANDALONE by default
 }
 
+// OnboardEndnodeWithGatewayRequest for requesting Onboard endnode with gateway id.
+type OnboardEndnodeWithGatewayRequest struct {
```
Comment (satoshikumano) [inline api.go]: **OnboardEndnodeWithGatewayRequest**
naming looks indecisive.
WithGateway doesn't adds much information to the struct but the makes the name redundant.

Suggested solutions:

#### Fully Qualified:

Prepare 2 structs for the different type of the content-type.
- OnboardEndnodeWithGatewayThingIDRequest
- OnboardEndnodeWithGatewayVendorThingIDRequest

#### Pack it in one struct.
- OnboardEndnodeRequest

The only difference between those content-types are gatewayVendorThingID and gatewayThingID.
Only accept either of them. (Somehow.)

File: api.go [hunk 2]
```diff
@@ -189,6 +189,22 @@ type OnboardByOwnerRequest struct {
...
+	EndNodeVendorThingID   string `json:"endNodeVendorThingID"`
+	EndNodePassword        string `json:"endNodePassword"`
+	GatewayThingID         string `json:"gatewayThingID"`
+	Owner                  string `json:"owner"`
+	EndNodeThingProperties string `json:"endNodeThingProperties,omitempty"`
+	EndNodeThingType       string `json:"endNodeThingType,omitempty"`
+}
+
+// OnboardEndnodeWithGatewayResponse for receiving response of onboarding endnode with gateway
+type OnboardEndnodeWithGatewayResponse struct {
```
Comment (satoshikumano) [inline api.go]: Response is common.
**OnboardEndnodeResponse** would be fine.

**Comment by satoshikumano** [comment]
@yongpingchen  please confirm.

File: api.go (hunk 1 above)
Comment (yongpingchen) [inline api.go]: Thanks @satoshikumano,  I prefer to use `OnboardEndnodeRequest`

**Comment by yongpingchen** [comment]
@satoshikumano could you confirm?

File: api.go [hunk 3]
```diff
@@ -189,6 +189,34 @@ type OnboardByOwnerRequest struct {
 	LayoutPosition string `json:"layoutPosition,omitempty"` // pattern: GATEWAY|STANDALONE|ENDNODE, STANDALONE by default
 }
 
+// OnboardEndnodeBasicRequest is the command fields for
+// OnboardEndnodeWithGatewayThingIDRequest and OnboardEndnodeWithGatewayVendorThingIDRequest
+type OnboardEndnodeBasicRequest struct {
```
Comment (satoshikumano) [inline api.go]: OnboardEndnodeRequestCommon

With 'Basic' it looks like GatewayThingID or GatewayVendorThingID is optional.

**Comment by yongpingchen** [comment]
thanks @satoshikumano I renamed it

File: author.go [hunk 4]
```diff
@@ -229,7 +231,40 @@ func (a *APIAuthor) OnboardThingByOwner(request OnboardByOwnerRequest) (*Onboard
...
+	if err := json.Unmarshal(bodyStr, &ret); err != nil {
+		return nil, err
+	}
+
+	return &ret, nil
+}
+
+// OnboardEndnodeWithGateway onboards an endnode
+// request must be either OnboardEndnodeWithGatewayVendorThingIDRequest or OnboardEndnodeWithGatewayThingIDRequest
+func (a *APIAuthor) OnboardEndnodeWithGateway(request interface{}) (*OnboardEndnodeResponse, error) {
```
Comment (satoshikumano) [inline author.go]: OnboardEndnodeWithGatewayVendorThingID
OnboardEndnodeWithGatewayThingID

**Pull request merged**